editoast/src/views/v2/path/projection.rs

```diff
@@ -0,0 +1,471 @@
+        Self {
+            track_section: track_section.as_ref().into(),
+            offset,
+        }
+    }
+}
+
+impl<'a> PathProjection<'a> {
+    /// Retrieve a track range from the path given the track section identifier.
+    fn get_track_range(&self, track_section: &Identifier) -> Option<&TrackRange> {
+        let index = *self.track_index.get(track_section)?;
+        Some(&self.path[index])
+    }
+
+    /// Create a new projection from a path.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the path contains multiple times the same track section.
+    pub fn new(path: &'a Vec<TrackRange>) -> Self {
+        let mut track_index = HashMap::new();
+        let mut map_position = HashMap::new();
+        let mut pos: u64 = 0;
+        for (index, track_range) in path.iter().enumerate() {
+            assert!(track_index
+                .insert(&track_range.track_section, index)
+                .is_none());
+            map_position.insert(&track_range.track_section, pos);
+            pos += track_range.end - track_range.begin;
+        }
+        Self {
+            path,
+            map_position,
+            track_index,
+            length: pos,
+        }
+    }
+
+    /// Get the position in the path given a location (track section and offset).
+    pub fn get_position(&self, location: &TrackLocation) -> Option<u64> {
+        let base_position = *self.map_position.get(&location.track_section)?;
+        let track_range = self
+            .get_track_range(&location.track_section)
+            .expect("Track range should be found since we found the base_position");
+        // Check if the offset is in the range.
+        if location.offset < track_range.begin || location.offset > track_range.end {
+            return None;
+        }
+        if track_range.direction == Direction::StartToStop {
+            Some(base_position + location.offset - track_range.begin)
+        } else {
+            Some(base_position + track_range.end - location.offset)
+        }
+    }
+
+    /// Get the location (track section and offset) given a position in a path.
+    /// Is function uses a binary search to find the track section and offset.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the position is out of bounds.
+    pub fn get_location(&self, position: u64) -> TrackLocationFromPath {
+        assert!(position <= self.length, "Position out of bounds");
+
+        // Binary search to retrieve the corresponding track section range.
+        let mut left = 0;
+        let mut right = self.path.len() - 1;
+        while left != right {
+            let mid = (left + right + 1) / 2;
+            let mid_track_range = &self.path[mid];
+            let mid_position = self.map_position[&mid_track_range.track_section];
+
+            if mid_position > position {
+                right = mid - 1;
+            } else {
+                left = mid;
+            }
+        }
```

Comment:
```suggestion
        let track_section_range_index = {
            let mut left = 0;
            let mut right = self.path.len() - 1;
            while left != right {
                let mid = (left + right + 1) / 2;
                let mid_track_range = &self.path[mid];
                let mid_position = self.map_position[&mid_track_range.track_section];

                if mid_position > position {
                    right = mid - 1;
                } else {
                    left = mid;
                }
            }
            left
        };
```

That's just style but for a function this size I think restricting scope as much as possible improves readability. (Or you can turn this into a local function if you prefer ofc.)